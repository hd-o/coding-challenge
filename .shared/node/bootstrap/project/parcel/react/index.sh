#!/usr/bin/env bash

project_name=$1

bash ../../.shared/node/bootstrap/project/parcel/index.sh $project_name
bash ../../.shared/node/bootstrap/react/index.sh
