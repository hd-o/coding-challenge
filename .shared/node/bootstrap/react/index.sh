#!/usr/bin/env bash

mkdir -p ./.config

cp ../.shared/node/bootstrap/react/config/eslint.js.txt ./.config/eslint.js

yarn add \
  @testing-library/jest-dom \
  @testing-library/react \
  @types/react \
  @types/react-dom \
  eslint-plugin-react \
  eslint-plugin-react-hooks \
  react \
  react-dom
