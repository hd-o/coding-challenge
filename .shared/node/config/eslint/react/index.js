// @ts-check

const base = require('../')
const { writeConfig } = require('../../write')

/**
 * @type {import('eslint').Linter.Config}
 */
const eslintrc = {
  ...base.eslintrc,
  extends: [
    ...base.eslintrc.extends,
    "plugin:react-hooks/recommended",
  ],
  plugins: [
    ...base.eslintrc.plugins,
    "react",
  ],
}

const writeEslintrc = (config = eslintrc) => writeConfig(config, './.eslintrc')

module.exports = { eslintrc, writeEslintrc }
