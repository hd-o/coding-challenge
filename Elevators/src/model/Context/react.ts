/**
 * @file
 * @see {@link ./ContextProvider.tsx}
 * Context used by React UI tree
 */
import { range } from 'lodash'
import { createContext } from 'react'
import { Elevator } from '../type/Elevator'
import { ElevatorController } from '../type/Elevator/Controller'
import { Floor } from '../type/Floor'
import { ModelContext } from './model'

const ctx = new ModelContext()
/** 6 Floors from 0 to 5 */
const floors = range(6).map<Floor>((number) => new Floor(number, ctx))
/** Context with 3 Elevators */
export const defaultValue = {
  elevatorService: new ElevatorController(
    range(3).map(() => new Elevator(floors, ctx)),
  ),
}

export const UiContext = createContext(defaultValue)
