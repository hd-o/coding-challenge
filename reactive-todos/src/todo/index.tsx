import { createContext, FC, useContext } from 'react'
import { TodoCtrlCtx } from './controller'
import { ITodo } from './model'

type TodoHandler = (t: ITodo) => () => void

interface Props {
  todo: ITodo
}

const Todo: FC<Props> = (props) => {
  const { todo } = props
  const todoCtrl = useContext(TodoCtrlCtx)()

  const handleToggle: TodoHandler = (todo) => {
    return () => todoCtrl.toggleComplete(todo.id)
  }

  const handleDelete: TodoHandler = (todo) => {
    return () => todoCtrl.delete(todo.id)
  }

  return (
    <li>
      <button
        onClick={handleToggle(todo)}
        title={`Toggle: ${todo.text}`}
      >
        {todo.done ? '🟢' : '⚪'}
      </button>
      {' '}
      <button
        onClick={handleDelete(todo)}
        title={`Delete: ${todo.text}`}
      >
        {'🗑️'}
      </button>
      {' '}
      <span style={{ textDecoration: todo.done ? 'line-through' : 'none' }}>
        {todo.text}
      </span>
    </li>
  )
}

export const TodoCtx = createContext(Todo)
