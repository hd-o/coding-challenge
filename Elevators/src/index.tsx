import 'reflect-metadata'
import './debug'
import { render } from 'react-dom'
import { App } from './component/App'
import { ContextProvider } from './model/Context/Provider'

render(
  <ContextProvider>
    <App />
  </ContextProvider>,
  document.getElementById('root'),
)
