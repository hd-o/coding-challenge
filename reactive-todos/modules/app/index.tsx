import { createContext, useContext } from 'react'
import { TodoInputCtx } from '../todo/input'
import { TodoListCtx } from '../todo/list'

function App() {
  const TodoInput = useContext(TodoInputCtx)
  const TodoList = useContext(TodoListCtx)
  return (
    <div style={{ padding: '5px 0 0 20px' }}>
      <h2>To-Dos</h2>
      <TodoInput />
      <TodoList />
    </div>
  )
}

export const AppCtx = createContext(App)
