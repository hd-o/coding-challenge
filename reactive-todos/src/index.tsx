import '../public/global.css'
import 'reflect-metadata'
import { render } from 'react-dom'
import { App } from './app'

render(<App />, document.getElementById('root'))
