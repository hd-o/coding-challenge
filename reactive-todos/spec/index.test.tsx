/**
 * @jest-environment jsdom
 */
import 'reflect-metadata'
import '@testing-library/jest-dom'
import { App } from '/src/app'
import { container } from 'tsyringe'
import { render, screen } from '@testing-library/react'
import { TestUtil } from './util'

let util: TestUtil

beforeEach(() => {
  container.clearInstances()
  util = container.resolve(TestUtil)
  render(<App />)
})

describe('To-Do App', () => {
  test('App snapshot', () => {
    const app = render(<App />)
    expect(app.asFragment()).toMatchSnapshot('app')
  })
  describe('Add To-Do', () => {
    test('to-do input submit', async () => {
      await util.submitTodo(util.mockValues[0])
      const todo = screen.getByText(util.mockValues[0])
      expect(todo).toBeDefined()
      expect(util.todoInput).toHaveTextContent('')
    })
    test('add to-do button', async () => {
      await util.addTodo(util.mockValues[0])
      const todo = screen.getByText(util.mockValues[0])
      expect(todo).toBeDefined()
      expect(util.todoInput).toHaveTextContent('')
    })
  })
  describe('Toggle To-Do', () => {
    test('toggle to-do', async () => {
      await util.addTodo(util.mockValues[0])
      await util.toggleTodo(util.mockValues[0])
      const toggled = screen.getByText('ðŸŸ¢')
      expect(toggled).toBeDefined()
    })
    test('to-do is toggled individually', async () => {
      await util.addTodo(util.mockValues[0])
      await util.addTodo(util.mockValues[1])
      await util.toggleTodo(util.mockValues[0])
      // getByText should find only 1 item
      const toggled = screen.getByText('ðŸŸ¢')
      expect(toggled).toBeDefined()
    })
  })
  describe('Delete To-Do', () => {
    test('delete to-do', async () => {
      await util.addTodo(util.mockValues[0])
      const todo = screen.getByText(util.mockValues[0])
      expect(todo).toBeDefined()
      await util.deleteTodo(util.mockValues[0])
      const todoNotFound = util.todoNotFound(util.mockValues[0])
      expect(todoNotFound).toBe(true)
    })
    test('todo should be deleted individually', async () => {
      await util.addTodo(util.mockValues[0])
      await util.addTodo(util.mockValues[1])
      await util.deleteTodo(util.mockValues[0])
      const todoNotFound = util.todoNotFound(util.mockValues[0])
      expect(todoNotFound).toBe(true)
      const todo2 = screen.getByText(util.mockValues[1])
      expect(todo2).toBeDefined()
    })
  })
})
