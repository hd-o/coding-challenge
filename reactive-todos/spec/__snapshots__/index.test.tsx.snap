// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`To-Do App App snapshot: app 1`] = `
<DocumentFragment>
  <div
    style="padding: 5px 0px 0px 20px;"
  >
    <h2>
      To-Dos
    </h2>
    <form>
      <input
        data-testid="todo-input"
        type="input"
        value=""
      />
      Â 
      <button
        data-testid="add-todo-button"
        style="cursor: pointer;"
        type="submit"
      >
        +
      </button>
    </form>
    <p>
      <code>
        Insert a To-Do through the input above
      </code>
    </p>
  </div>
</DocumentFragment>
`;
